src/message_bus.zig

```diff
@@ -686,7 +715,7 @@ const Connection = struct {
 
         const bytes_received = result catch |err| {
             // TODO: maybe don't need to close on *every* error
-            log.err("error receiving body from {}: {}", .{ self.peer, err });
+            log.err("error receiving from {}: {}", .{ self.peer, err });
             self.shutdown();
```

Comment:
Hmm, there could be a send currently in progress as well but I'm not sure if the shutdown syscall is worth it in that case or not. It could potentially take an unbounded amount of time to complete if the send buffer for the socket in the kernel is full, so I was inclined to issue the `shutdown()` syscall in this case to be on the safe side.

Moving this logic into our `shutdown()` function does sound like it could be a nice cleanup though and prevent misuse.

Reply:
Agreed, let's err on the side of doing a shutdown syscall within `shutdown()` if there are any pending `recv`s or `send`s, let's make this logic internal to `shutdown()` and then let's replace our `maybe_close()` code paths with a single `shutdown()` code path so that we only set `.state = shutting_down` from within `shutdown()`.

Reply:
It turns out we can't make the logic internal to shutdown as there are cases where we don't want to make the shutdown syscall despite a send or recv being in progress (e.g. when the peer closes their side of the connection and we recv 0 bytes). 

I was still able to clean things up a bit in 691e089 however.